### DISCUSSION TITLE
Generating a new pass from frontend overwrites existing pass in wallet instead of creating a new one

### DISCUSSION THREAD

**Discussion by mbappai** [alexandercerutti/passkit-generator#140] [Q&A]
Hey Alex! I don't know why whenever I generate a new pass with different props every time from frontend, instead of add that newly generated pass to my apple wallet it just end up overwriting the existing one. Please help. Thanks

**Comment by alexandercerutti** [comment]
Hey @mbappai do you mind sharing your code?

**Reply by mbappai** [reply]
sure! here it is
```

import {PKPass} from 'passkit-generator'

import getCertificates from './getCertificates'
import path from 'path'
import { NextApiResponse } from 'next'
import dayjs from 'dayjs'
import exp from 'constants'

export default async function handler(req:any, res:NextApiResponse){

    const body = JSON.parse(req?.body)

    const {qrCode,expiryDate, quantity, targetDate, eventName, price, street, location} = body

    const {signerCert, signerKey, wwdr, signerKeyPassphrase} = await getCertificates()
    const pass = await PKPass.from({
        model: path.resolve(
            // __dirname,
            // "../../../../",
            './ticket.pass',
        ),
        certificates:{ 
            wwdr,
            signerKey,
            signerCert,
            signerKeyPassphrase
        }
    })

    // pass.set
    // pass.headerFields.push(
    //     {
    //         key: "header1",
    //         label: "Event",
    //         value: "Flexable natural line skips",
    //         textAlignment: "PKTextAlignmentCenter",
    //     },
    //     {
    //         key: "header2",
    //         label: "Description",
    //         value: "Best service in Syracuse",
    //         textAlignment: "PKTextAlignmentCenter",
    //     },
    // );

    pass.setLocations(location)

    pass.setBarcodes({
        message: JSON.stringify(qrCode),
        format: "PKBarcodeFormatQR",
        altText: "Qr code for bars"
    })


    pass.secondaryFields.push({

            "key": "location",
            "label": "LOCATION",
            "value": street
        
    })

    pass.auxiliaryFields.push(
        {
            "key": "quantity",
            "value": `${quantity} Ticket(s)`,
            "label":'Quantity',
            "row": 0
        },
        {
            "key": "price",
            "value": `$${price}`,
            "label":'Price',
            "row": 0
        },
        {
            "key": "validOn",
            "value": dayjs(targetDate).format("MMM DD, YYYY"), //  convert this to us timezone
            "label":'Valid On',
            "row": 0
        },
    )
    

    pass.primaryFields.push({
        key: "Name",
        label: "Service Name",
        value: eventName,
        textAlignment: "PKTextAlignmentLeft",
    }) 
 
    pass.setExpirationDate(new Date(expiryDate))

    pass.setRelevantDate(new Date(targetDate))

    // pass.type = "eventTicket"
    
    const buffer = pass.getAsBuffer();

    // console.log(pass.props['barcodes'])
    
    // console.log('streaming...',buffer)
    res.status(200) 
    res.setHeader("Content-Type", pass.mimeType)
    res.send(buffer)

}  






```

**Reply by alexandercerutti** [reply]
Where do you set the `serialNumber`?

**Reply by mbappai** [reply]
ohh damn! It has a static value in pass.json

**Reply by alexandercerutti** [reply]
Attempt to make it dynamic. Let me know 😄

**Reply by mbappai** [reply]
hahaha alright! will do. Thanks man

**Reply by mbappai** [reply]
will this expression be valid?

`pass.props['serialNumber'] = uniqueValue`

**Reply by alexandercerutti** [reply]
Nope, as per [the `.props` documentation](https://github.com/alexandercerutti/passkit-generator/wiki/API-Documentation-Reference#props-getter-only), it will return a deep copy of the props.

You need to set the props in the constructor.
Look at it here: https://github.com/alexandercerutti/passkit-generator/wiki/API-Documentation-Reference#-from-a-template

**Reply by mbappai** [reply]
oh so like this then

```
   const pass = await PKPass.from({
        model: path.resolve(
            // __dirname,
            // "../../../../",
            './ticket.pass',
        ),
        certificates:{ 
            wwdr,
            signerKey,
            signerCert,
            signerKeyPassphrase
        },
        serialNumber: ticketSecret
    })

```

**Reply by alexandercerutti** [reply]
No, the second parameter is still an object. Look how is used `getAdditionalPropsForThisPassSomehow`.

**Reply by mbappai** [reply]
Like this then. Sorry a bit hard for me to grasp seeing as `getAdditionalPropsSomehow` function doesn't show a return type or anything.

```
   const pass = await PKPass.from({
        model: path.resolve(
            // __dirname,
            // "../../../../",
            './ticket.pass',
        ),
        certificates:{ 
            wwdr,
            signerKey,
            signerCert,
            signerKeyPassphrase
        }
    },
   {
              serialNumber: ticketSecret
    }
)

```

**Reply by alexandercerutti** [reply]
Exactly.
I actually hoped and assumed that the description of `getAdditionalPropsForThisPassSomehow`, as is it showed [above the example](https://github.com/alexandercerutti/passkit-generator/wiki/API-Documentation-Reference#creating-a-pass-from-a-different-source-like), was enough.

Anyway, as per the fact your are calling it "ticketSecret", be sure it is not actually a secret that can be used for something bad, because passes can get downloaded (if not told otherwise) and logged in Console.app

**Reply by mbappai** [reply]
haha you're right it was enough. My bad for not seeing that part of the docs. Let me try than out and see how it goes

**Reply by alexandercerutti** [reply]
P.s. also the example in the readme shows probably a better example ahah

**Reply by mbappai** [reply]
Mahn! props to you for all your hardwork alex! One would think writing the lib will be the end of project, but documentation itself takes a lot of discipline and patience to get right, which you nailed!

**Reply by alexandercerutti** [reply]
Now I honestly expect a donation 😂

Yeah, I know, the documentation is one of the strong points of this library and I'm attempting to follow the same "pattern" with all the other libraries I'm creating.

**Reply by mbappai** [reply]
hahaah totally man! You deserve every penny!

What other libs are you working on right now? Do tell

**Reply by alexandercerutti** [reply]
Published this some weeks ago: https://github.com/alexandercerutti/sub37, which is waaaaay bigger than passkit-generator 😂 like a year of development, and it is not even completed yet.

**Reply by mbappai** [reply]
Damn!! I think I did see a post you made on linkedin regarding you working on this project. Thought it was a project for your current place of work

**Reply by alexandercerutti** [reply]
Yeah, probably. The idea came while working on the current work project, but I created it on my own.

**Reply by mbappai** [reply]
I'm certainly going to be watching that project and try to help out where ever I can
